veracruz-utils/src/csr.rs

```diff
@@ -0,0 +1,293 @@
+    let overall_length:u16 = (template.overall_length_initial_value + signature_length as u16) as u16;
+    if overall_length < 256 {
+        constructed_csr[template.overall_length_field_location.0] = overall_length as u8;
+    } else {
+        constructed_csr[template.overall_length_field_location.0] = ((overall_length & 0xff00) >> 8) as u8;
+        constructed_csr[template.overall_length_field_location.0 + 1] = (overall_length & 0xff) as u8;
+    }
+
+    return Ok(constructed_csr.clone());
+}
+
+const CSR_PUBKEY_LOCATION: (usize, usize) = (129 + 26, 220);
+pub fn convert_csr_to_cert(csr: &[u8], cert_template: &CertTemplate, enclave_hash: &[u8], private_key: &EcdsaKeyPair) -> Result<std::vec::Vec<u8>, CertError> {
+    let mut constructed_cert = cert_template.template.to_vec();
+    let valid_from = generate_utc_time(2021, 5, 2, 17, 1, 0)?;
+    constructed_cert.splice(cert_template.valid_from_location.0..cert_template.valid_from_location.1,
+        valid_from,
+    );
+    // TODO: Once the root enclave is gone, this can be done properly inside the proxy service
+    let valid_until = generate_utc_time(2031, 5, 30, 17, 1, 0)?;
+    constructed_cert.splice(cert_template.valid_until_location.0..cert_template.valid_until_location.1,
+        valid_until,
+    );
+
+    // replace the public key in the template with the public key from the CSR
+    let public_key = &csr[CSR_PUBKEY_LOCATION.0..CSR_PUBKEY_LOCATION.1];
+    constructed_cert.splice(cert_template.public_key_location.0..cert_template.public_key_location.1, public_key.to_vec());
+
+    // replace the dummy data in the template with the enclave hash
+    constructed_cert.splice(cert_template.enclave_hash_location.0..cert_template.enclave_hash_location.1, enclave_hash.to_vec());
+
+    // Sign the body of the constructed cert 
+    let signature: Vec<u8> = private_key.sign(&SystemRandom::new(),
+                                              &constructed_cert[cert_template.signature_range.0..cert_template.signature_range.1]).unwrap().as_ref().to_vec();
+    // place the signature in the constructed_cert
+    let signature_length = signature.len();
+    constructed_cert.splice(
+        cert_template.signature_location.0..cert_template.signature_location.1,
+        signature,
+    );
+
+    let signature_field_length:u8 = (cert_template.signature_length_initial_value + signature_length as u8) as u8;
+    constructed_cert[cert_template.signature_length_field_location.0] = signature_field_length;
+
+    let overall_length:u16 = (cert_template.overall_length_initial_value + signature_length as u16) as u16;
+    constructed_cert[cert_template.overall_length_field_location.0] = ((overall_length & 0xff00) >> 8) as u8;
+    constructed_cert[cert_template.overall_length_field_location.0 + 1] = (overall_length & 0xff) as u8;
+
+    return Ok(constructed_cert.clone());
+}
+
+pub fn generate_utc_time(year: u32, month: u8, day: u8, hour: u8, minute: u8, second: u8) -> Result<Vec<u8>, CertError> {
```

Comment:
It's maybe easier on the caller if instead of `u8` values you take a `T` with constraint `T : Into<u8> + Clone`.

Reply:
I've been trying to get this to work for a bit, and it might be easier on the caller, but it does complicate the code a bit. 

Since the function uses a format!(), it also needs to support the constraint `Display`. Additionally, the `CertError::InvalidUtcInputs` can't take `T` as input. It wants the input to match it's types exactly (and I don't think you can add constraints to an enum. 

I'm certainly willing to look at this further, but it's not a trivial change. But I would learn a lot doing it :-)

Reply:
Nevermind. I think I figured it out.